File: TwoPointer/76.minimum-window-substring.go
package leetcode

import "math"

// @leet start

// use a acount to check if all covered
// func allCovered(freq map[rune]int) bool {
// 	for _, v := range freq {
// 		if v > 0 {
// 			return false
// 		}
// 	}
// 	return true
// }

func minWindow(s string, t string) string {
	freq := map[rune]int{}
	for _, r := range t {
		freq[r]++
	}
	var bestL, bestR, l, r int
	minLen, count := math.MaxInt, len(t)
	for ; r < len(s); r++ {
		curR := rune(s[r])
		// move r until all covered
		if _, ok := freq[curR]; !ok {
			continue
		} else {
			freq[curR]--
			if freq[curR] >= 0 {
				count--
			}
			if count > 0 {
				continue
			}
			// if !allCovered(freq) {
			// 	continue
			// }
		}

		// allCovered, try to shirk l side
		for {
			lrune := rune(s[l])
			if _, ok := freq[lrune]; !ok {
				l++
			} else {
				if freq[lrune]+1 > 0 {
					break
				}
				freq[lrune]++
				l++
			}
		}

		curLen := r - l + 1
		if curLen < minLen { // need <=
			minLen = curLen
			bestL, bestR = l, r
		}
	}
	if minLen == math.MaxInt {
		return ""
	}
	return s[bestL : bestR+1]
}

// @leet end
// original approach
func allCovered(freq map[rune]int) bool {
	for _, v := range freq {
		if v > 0 {
			return false
		}
	}
	return true
}

func minWindow2(s string, t string) string {
	freq := map[rune]int{}
	for _, r := range t {
		freq[r]++
	}
	var bestL, bestR, l, r int
	minLen := len(s)
	for ; r < len(s); r++ {
		curR := rune(s[r])
		// move r until all covered
		if _, ok := freq[curR]; !ok {
			continue
		} else {
			freq[curR]--
			if !allCovered(freq) {
				continue
			}
		}

		// allCovered, try to shirk l side
		for {
			lrune := rune(s[l])
			if _, ok := freq[lrune]; !ok {
				l++
			} else {
				if freq[lrune]+1 > 0 {
					break
				}
				freq[lrune]++
				l++
			}
		}

		curLen := r - l + 1
		if curLen <= minLen { // need <=
			minLen = curLen
			bestL, bestR = l, r
		}
	}
	if !allCovered(freq) {
		return ""
	}
	return s[bestL : bestR+1]
}

// use array
func minWindow(s string, t string) string {
	freq := [128]int{}    // can directly use array as char freq
	for _, r := range t { // up vote
		freq[r]++
	}
	var bestL, bestR, l, r int
	minLen, count := math.MaxInt, len(t)
	for ; r < len(s); r++ {
		curR := s[r]
		// move r until all covered
		freq[curR]--
		if freq[curR] >= 0 {
			count--
		}
		// if count > 0 {
		// 	continue
		// } // logic can move to next

		// if !allCovered(freq) {
		// 	continue
		// }

		// allCovered, try to shirk l side
		// for {
		for count == 0 {
			curLen := r - l + 1
			if curLen < minLen { // need <=
				minLen = curLen
				bestL, bestR = l, r
			}
			lrune := s[l]
			// if freq[lrune]+1 > 0 {
			//     count++
			// 	// break
			// }
			freq[lrune]++
			l++
			if freq[lrune] > 0 {
				count++
			}
		}

	}
	if minLen == math.MaxInt {
		return ""
	}
	return s[bestL : bestR+1]
}

// use a acount to check if all covered
// func allCovered(freq map[rune]int) bool {
// 	for _, v := range freq {
// 		if v > 0 {
// 			return false
// 		}
// 	}
// 	return true
// }

func minWindow(s string, t string) string {
	var (
		freq          [128]int
		l, bestl      int
		minLen, count = -1, len(t)
	)
	for _, b := range t {
		freq[b]++
	}
	for r, b := range s {
		freq[b]--
		if freq[b] >= 0 {
			count--
		}
		for count == 0 { // all freq [t] <= 0
			curLen := r - l + 1
			if minLen == -1 || curLen < minLen { // can be <= minLen does not matter
				minLen, bestl = curLen, l
			}
			lb := s[l] // pop left byte
			freq[lb]++
			l++
			if freq[lb] > 0 { // each byte is a single trigger point
				count++
			}
		}

	}
	if minLen == -1 {
		return ""
	}

	return s[bestl : bestl+minLen]
}

File: TwoPointer/76.minimum-window-substring.py
# @leet start
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        freq = [0] * 128
        count = len(t)
        minLen = bestL = l = 0
        for c in t:
            freq[ord(c)] += 1
        ## go over s
        for r, c in enumerate(s):
            freq[ord(c)] -= 1  # NOTE: can be negative even for freq count
            # NOTE: freq like presSUm, always update.count is different
            if freq[ord(c)] >= 0:  # make sure each site is 0
                count -= 1

            while count == 0:
                curL = r - l + 1
                if minLen == 0 or curL < minLen:
                    minLen = curL
                    bestL = l
                l_c = s[l]
                l += 1
                freq[ord(l_c)] += 1  # NOTE: + not -
                if freq[ord(l_c)] > 0:
                    count += 1

        return s[bestL : bestL + minLen]


# @leet end
def minWindow(self, s: str, t: str) -> str:
        window = Counter()
        need = Counter(t)
        l = r = minL = bestL = valid = 0
        for r in range(len(s)):
            c = s[r]

            if need[c]:
                window[c] += 1
                if window[c] == need[c]:
                    valid += 1

            while valid == len(
                need
            ):  # NOTE: can include other char, no need fix window
                curL = r - l + 1
                if minL == 0 or curL < minL:
                    minL = curL
                    bestL = l
                d = s[l]
                l += 1

                if d in need:
                    window[d] -= 1
                    if window[d] < need[d]:
                        valid -= 1

        return s[bestL : bestL + minL]
